testcases/crypto/dilithium_func.c

```diff
@@ -0,0 +1,297 @@
+        if (rc == CKR_MECHANISM_INVALID) {
+            /* no support for EC key gen? skip */
+            testcase_skip("Slot %u doesn't support CKM_IBM_DILITHIUM ",
+                          (unsigned int) SLOT_ID);
+            rc = CKR_OK;
+            goto testcase_cleanup;
+        } else {
+            testcase_error("C_GetMechanismInfo() rc = %s", p11_get_ckr(rc));
+            goto testcase_cleanup;
+        }
+    }
+
+    /* Setup attributes for public/private Dilithium key */
+    CK_BBOOL attr_sign = TRUE;
+    CK_BBOOL attr_verify = TRUE;
+    CK_ATTRIBUTE dilithium_attr_private[] = {
+        {CKA_SIGN, &attr_sign, sizeof(CK_BBOOL)},
+    };
+    CK_ATTRIBUTE dilithium_attr_public[] = {
+        {CKA_VERIFY, &attr_verify, sizeof(CK_BBOOL)},
+    };
+
+    /* Generate Dilithium key pair */
+    rc = funcs->C_GenerateKeyPair(session, &mech,
+                   dilithium_attr_public, 1,
+                   dilithium_attr_private, 1,
+                   &publ_key, &priv_key);
+    testcase_new_assertion();
+    if (rc != CKR_OK) {
+        if (rc == CKR_MECHANISM_PARAM_INVALID ||
+            rc == CKR_ATTRIBUTE_VALUE_INVALID) {
+            testcase_skip("Slot %u doesn't support Dilithium",
+                          (unsigned int) SLOT_ID);
+            goto testcase_cleanup;
+        }
+        testcase_fail
+            ("C_GenerateKeyPair with valid input failed, rc=%s",
+             p11_get_ckr(rc));
+        goto testcase_cleanup;
+    }
+    testcase_pass("*Generate Dilithium key pair passed.");
+
+    /* Sign/verify with this key pair */
+    for (j = 0; j < (sizeof(signVerifyInput) / sizeof(_signVerifyParam)); j++) {
+        testcase_new_assertion();
+        rc = run_SignVerifyDilithium(session,
+                               signVerifyInput[j].mechtype,
+                               signVerifyInput[j].inputlen,
+                               priv_key, publ_key);
+        if (rc != 0) {
+            testcase_fail("run_SignVerifyDilithium failed index=%lu.", j);
+            goto testcase_cleanup;
+        }
+        testcase_pass("*Sign & verify j=%lu passed.", j);
+    }
+
+    rc = CKR_OK;
+
+testcase_cleanup:
+    if (publ_key != CK_INVALID_HANDLE)
+        funcs->C_DestroyObject(session, publ_key);
+    if (priv_key != CK_INVALID_HANDLE)
+        funcs->C_DestroyObject(session, priv_key);
+
+    testcase_close_session();
+
+    return rc;
+}
+
+int main(int argc, char **argv)
+{
+    CK_C_INITIALIZE_ARGS cinit_args;
+    int rc;
+    CK_RV rv;
+
+    rc = do_ParseArgs(argc, argv);
+    if (rc != 1)
+        return rc;
+
+    printf("Using slot #%lu...\n\n", SLOT_ID);
+    printf("With option: no_init: %d\n", no_init);
+
+    rc = do_GetFunctionList();
+    if (!rc) {
+        PRINT_ERR("ERROR do_GetFunctionList() Failed , rc = 0x%0x\n", rc);
+        return rc;
+    }
+
+    memset(&cinit_args, 0x0, sizeof(cinit_args));
+    cinit_args.flags = CKF_OS_LOCKING_OK;
+
+    funcs->C_Initialize(&cinit_args);
+
+    {
```

Comment:
Whats the idea behind checking for the correct GetFunctionState resp CancelFunction return values in the dilithium test case ?

Reply:
Good point. All of our opencryptoki testcases, without any exception, show this same piece of code. Dilithium is just the last one adapting this. When looking at the implementations of C_GetFunctionStatus and C_CancelFunction, it appears that they check if opencryptoki is initialized and would return CKR_CRYPTOKI_NOT_INITIALIZED if it's not. So this check just makes sure that it's ok to start with the testcase, probably to avoid strange errors later.

Reply:
To make this more clear, we could introduce one common macro, like CHECK_CRYPTOKI_INITIALIZED and use this in all testcases.